### ISSUE TITLE
Upsampling network may be simplified

### ISSUE THREAD

**Issue by geneing** [fatchord/WaveRNN#27]
I think the upsampling network can be replaced by a single torch.nn.Upsample operation with scale_factor equal to hop_length and mode='linear'. 

I took a network trained on LJSpeech data and looked at the output of upsampling layer. Upsampled mels from the upsampling network match (up to an arbitrary scaling factor) very closely the values I get by linearly interpolating between the original mel values. 

Auxiliary network is still needed, though.

**Comment by fatchord** [comment]
@geneing I probably would've agreed had someone not mentioned to me that my upsampling "is basically a gaussian convolution with a time-shift". So I had a look at the kernel weights after training and this is what I found: 
![screenshot_weights](https://user-images.githubusercontent.com/37295764/49430898-1becc880-f7a4-11e8-9139-958508bfc1b4.png)

That first kernel is going to be more important overall and I reckon that does indeed look something like a guassian but shifted to the right a bit. I checked out another model and found more or less the same thing. 

What do you make of it?

**Comment by bliep** [comment]
@fatchord here's the kernel reponses and the total impulse response of the upsampling layers (my kernels are a bit smoother, dunno why):
![screenshot 2018-12-04 at 15 29 44](https://user-images.githubusercontent.com/1833211/49448696-b9aeba80-f7d9-11e8-878e-68049abeecf4.png)

**Comment by bliep** [comment]
btw, linear interpolation is just convolution with a 'triangular' filter so linear upsampling might indeed be just as ok 👍

**Comment by hdmjdp** [comment]
@geneing so you juse use model="linear"?

**Comment by fatchord** [comment]
@geneing Maybe I'm totally off the mark here but hear me out... the fact that the filter shape is shifted forward in time intrigues me. I mean - what if one tried shifting the conditioning features back in time by the same amount as the offset in the the guassian? Would the guassian shape end up centred? Could one interpret the offset in such a manner as to say 'this model is prioritising future conditioning features over current? Why not have two upsampling networks and give both a slice of current and future conditioning features?

Sorry, that's a lot of questions but I'm just curious what you'd think about that line of reasoning.

**Comment by G-Wang** [comment]
@geneing , @fatchord Amazon recently had a paper on implementing a universal neural vocoder here: https://arxiv.org/abs/1811.06292.

their architecture is quite simple, unidirectional rnn followed by dense layer then softmax, conditioned on outputs from an up-sampling network (they use rnns for upsampling). From reading the paper it seems the most important thing to have is a variety of dataset (multi speaker, different languages, etc).
![screenshot from 2019-01-14 12-51-34](https://user-images.githubusercontent.com/14854468/51140448-6fb6ec80-17fb-11e9-9932-6acb83d10321.png)
